File: backend/internal/user/validate.go
package user

import (
	"errors"
	"regexp"
)

// ========================
// 正则表达式定义
// ========================

// 邮箱：标准格式 local@domain.tld，长度不超过 100（数据库 VARCHAR(100)）
var emailRegex = regexp.MustCompile(`^[a-zA-Z0-9._%+\-]+@[a-zA-Z0-9.\-]+\.[a-zA-Z]{2,}$`)

// 用户名：2-20 个字符，允许中文、英文字母、数字、下划线
var usernameRegex = regexp.MustCompile(`^[\p{Han}a-zA-Z0-9_]{2,20}$`)

// 密码 — 合法字符：英文字母、数字、常见特殊符号
var passwordCharsRegex = regexp.MustCompile("^[a-zA-Z0-9!@#$%^&*()\\-_=+\\[\\]{};:'\",.<>?/\\\\|`~]+$")

// 密码 — 至少包含一个英文字母
var passwordLetterRegex = regexp.MustCompile(`[a-zA-Z]`)

// 密码 — 至少包含一个数字
var passwordDigitRegex = regexp.MustCompile(`[0-9]`)

// ========================
// 校验错误定义
// ========================

var (
	ErrEmailTooLong       = errors.New("邮箱长度不能超过 100 个字符")
	ErrEmailFormat        = errors.New("邮箱格式不正确")
	ErrUsernameFormat     = errors.New("用户名只能包含中文、英文字母、数字和下划线，长度 2-20")
	ErrPasswordLength     = errors.New("密码长度必须为 8-72 个字符")
	ErrPasswordChars      = errors.New("密码包含不允许的字符")
	ErrPasswordNeedLetter = errors.New("密码必须包含至少一个英文字母")
	ErrPasswordNeedDigit  = errors.New("密码必须包含至少一个数字")
)

// ========================
// 校验函数
// ========================

// ValidateEmail 校验邮箱格式：标准 local@domain.tld，长度 ≤ 100
func ValidateEmail(email string) error {
	if len(email) > 100 {
		return ErrEmailTooLong
	}
	if !emailRegex.MatchString(email) {
		return ErrEmailFormat
	}
	return nil
}

// ValidateUsername 校验用户名格式：中文/英文/数字/下划线，2-20 字符
func ValidateUsername(username string) error {
	if !usernameRegex.MatchString(username) {
		return ErrUsernameFormat
	}
	return nil
}

// ValidatePassword 校验密码格式：8-72 字符，至少含一个字母和一个数字，仅允许合法字符
func ValidatePassword(password string) error {
	if len(password) < 8 || len(password) > 72 {
		return ErrPasswordLength
	}
	if !passwordCharsRegex.MatchString(password) {
		return ErrPasswordChars
	}
	if !passwordLetterRegex.MatchString(password) {
		return ErrPasswordNeedLetter
	}
	if !passwordDigitRegex.MatchString(password) {
		return ErrPasswordNeedDigit
	}
	return nil
}

File: frontend/src/utils/validate.ts
// ========================
// 前端输入校验 —— 正则表达式与规则说明
// 与后端 backend/internal/user/validate.go 保持一致
// ========================

// ---------- 邮箱 ----------
// 规则：标准格式 local@domain.tld，长度不超过 100 个字符
export const EMAIL_MAX_LENGTH = 100
export const EMAIL_REGEX = /^[a-zA-Z0-9._%+\-]+@[a-zA-Z0-9.\-]+\.[a-zA-Z]{2,}$/

// ---------- 用户名 ----------
// 规则：2-20 个字符，仅允许中文、英文字母、数字、下划线
export const USERNAME_REGEX = /^[\u4e00-\u9fa5a-zA-Z0-9_]{2,20}$/

// ---------- 密码 ----------
// 规则：
//   1. 长度 8-72 个字符（bcrypt 上限 72 字节）
//   2. 仅允许英文字母、数字和常见特殊符号 !@#$%^&*()-_=+[]{};:'",.<>?/\|`~
//   3. 必须包含至少一个英文字母
//   4. 必须包含至少一个数字
export const PASSWORD_MIN_LENGTH = 8
export const PASSWORD_MAX_LENGTH = 72
export const PASSWORD_CHARS_REGEX = /^[a-zA-Z0-9!@#$%^&*()\-_=+\[\]{};:'",.<>?/\\|`~]+$/
export const PASSWORD_LETTER_REGEX = /[a-zA-Z]/
export const PASSWORD_DIGIT_REGEX = /[0-9]/

// ========================
// 校验函数
// ========================

export function validateEmail(email: string): string | null {
  if (email.length > EMAIL_MAX_LENGTH) {
    return '邮箱长度不能超过 100 个字符'
  }
  if (!EMAIL_REGEX.test(email)) {
    return '邮箱格式不正确'
  }
  return null
}

export function validateUsername(username: string): string | null {
  if (!USERNAME_REGEX.test(username)) {
    return '用户名只能包含中文、英文字母、数字和下划线，长度 2-20'
  }
  return null
}

export function validatePassword(password: string): string | null {
  if (password.length < PASSWORD_MIN_LENGTH || password.length > PASSWORD_MAX_LENGTH) {
    return '密码长度必须为 8-72 个字符'
  }
  if (!PASSWORD_CHARS_REGEX.test(password)) {
    return '密码包含不允许的字符'
  }
  if (!PASSWORD_LETTER_REGEX.test(password)) {
    return '密码必须包含至少一个英文字母'
  }
  if (!PASSWORD_DIGIT_REGEX.test(password)) {
    return '密码必须包含至少一个数字'
  }
  return null
}
